### PR TITLE
Fix USING clause for JOINs

### LINKED ISSUE #2311: Fix USING clause for JOINs
Columns that appear in a `USING` clause should not require qualification. This is particularly useful for outer joins, because the non-qualified reference always references the non-null value. It is not clear if qualifying reference is allowed: PostgreSQL allows it but Oracle does not. Allowing it is useful in order to do `IS NULL` checks for outer joins, especially when no other columns are available.

```
presto> select x from (values (1)) a (x) join (values (1)) b (x) using (x);
Query 20150202_160415_12625_wu6ji failed: Column 'x' is ambiguous
```

Additionally, because `USING` makes them the same column, the column should only appear once in the output descriptor:

```
presto> select * from (values (1)) a (x) join (values (1)) b (x) using (x);
 x | x 
---+---
 1 | 1 
(1 row)
```

Correct behavior for the above:

```
postgresql=# select x from (values (1)) a (x) join (values (1)) b (x) using (x);
 x 

---
 1
(1 row)

postgresql=# select * from (values (1)) a (x) join (values (1)) b (x) using (x);
 x 

---
 1
(1 row)
```

Additional test cases that should return the value from the outside of the join:

``` sql
SELECT x FROM (VALUES (1), (2)) a (x) LEFT JOIN (VALUES (1)) b (x) USING (x);
SELECT x FROM (VALUES (1)) a (x) RIGHT JOIN (VALUES (1), (2)) b (x) USING (x);
```

### PR DESCRIPTION
**Pull request by yuananf**
Fix #2311

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -1178,12 +1178,32 @@ public void testJoinUsing()
...
+                "VALUES (1)"
+        );
+
+        assertQuery(
+                "SELECT x FROM (VALUES (1)) a (x) JOIN (VALUES (1)) b (x) USING (x)",
+                "VALUES (1)"
+        );
+
+        assertQuery(
+                "SELECT x, sum(y) FROM (VALUES (1, 2), (1, 3)) a (x, y) JOIN (VALUES (1, 2), (1, 3)) b (x, y) USING (x, y) GROUP BY x ORDER BY x",
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Please add newlines to make the query easier to read

**Comment by electrum** [comment]
Thanks for fixing this! Two more test cases:
```
SELECT x FROM (VALUES (1), (2)) a (x) LEFT JOIN (VALUES (1)) b (x) USING (x);
SELECT x FROM (VALUES (1)) a (x) RIGHT JOIN (VALUES (1), (2)) b (x) USING (x);
```
Both should return two rows with the values from the outer side of the join (`1, 2`). I verified them in PostgreSQL.

**Comment by yuananf** [comment]
Oh, I missed something here, the output column should always be the join's base table.
Will fix it.

**Comment by yuananf** [comment]
@electrum comments addressed.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -1178,12 +1178,54 @@ public void testJoinUsing()
     }
 
     @Test
+    public void testJoinUsingOutputColumn()
+            throws Exception
+    {
+        assertQuery(
+                "SELECT * FROM (VALUES (1)) a (x) JOIN (VALUES (1)) b (x) USING (x)",
+                "VALUES (1)"
+        );
```
Comment (electrum) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Code style: put the closing `);` on the previous line

**Comment by haozhun** [comment]
@yuananf Please add this test

    SELECT x FROM (VALUES (1), (2)) a (x) FULL JOIN (VALUES (2), (3)) b (x) USING (x);

And make sure it returns
```
 x
---
 1
 2
 3
(3 rows)
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -283,11 +285,18 @@ protected Type visitQualifiedNameReference(QualifiedNameReference node, Analysis
...
+                    && equivalentJoinFields.get(matches.get(1)) == null)) {
                 throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node.getName());
             }
 
-            Field field = Iterables.getOnlyElement(matches);
+            // if the field appears in join using, then find the corresponding base table's output field
+            Field field = matches.get(0);
+            if (matches.size() == 2 && equivalentJoinFields.get(matches.get(1)) == matches.get(0)) {
+                field = matches.get(1);
+            }
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: The logic here is not quite straighforward. The first `if` does not check that one of `equivalentJoinFields.get(matches.get(1)) == matches.get(0)`, `(equivalentJoinFields.get(matches.get(0)) == matches.get(1))` is true. But the second `if` assumes so.

I would recommend:
```
Field field = null;
if (matches.size() == 1) {
    field = Iterables.getOnlyElement(matches);
}
else if (matches.size() == 2) {
    if (equivalentJoinFields.get(matches.get(1)) == matches.get(0)) {
        field = matches.get(1);
    }
    else if (equivalentJoinFields.get(matches.get(0)) == matches.get(1)) {
        field = matches.get(0);
    }
}
if (field == null) {
    throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node.getName());
}
```

**Comment by yuananf** [comment]
@haozhun @electrum 
My understanding is the following queries should return same result

Full join
```SELECT x FROM (VALUES (1), (2)) a (x) FULL JOIN (VALUES (2), (3)) b (x) USING (x);```
```SELECT COALESCE(a.x, b.x) FROM (VALUES (1), (2)) a (x) FULL JOIN (VALUES (2), (3)) b (x) USING (x);```

Left join
```SELECT x FROM (VALUES (1), (2)) a (x) LEFT JOIN (VALUES (2), (3)) b (x) USING (x);```
```SELECT COALESCE(a.x, b.x) FROM (VALUES (1), (2)) a (x) LEFT JOIN (VALUES (2), (3)) b (x) USING (x);```

Right join
```SELECT x FROM (VALUES (1), (2)) a (x) RIGHT JOIN (VALUES (2), (3)) b (x) USING (x);```
```SELECT COALESCE(a.x, b.x) FROM (VALUES (1), (2)) a (x) RIGHT JOIN (VALUES (2), (3)) b (x) USING (x);```

But the ```SELECT *``` should return both ```a.x``` and ```b.x```
Correct me if I'm wrong.

**Comment by haozhun** [comment]
You understanding of `SELECT x` is correct. Additionally, as you are aware of, `COALESCE` in left join and right join can be simplified.

`SELECT *` should produce the same output as `SELECT x` in the query example you given. There should be only 1 column. You can confirm this by running the query in postgres or oracle. We also looked at spec to confirm this.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 4]
```diff
@@ -1243,17 +1243,58 @@ public void testJoinUsing()
...
+                "SELECT * FROM (VALUES (1)) a (x) JOIN (VALUES (1)) b (x) USING (x)",
+                "VALUES (1)");
+
+        assertQuery(
+                "SELECT x FROM (VALUES (1)) a (x) JOIN (VALUES (1)) b (x) USING (x)",
+                "VALUES (1)");
+
+        assertQuery(
+                "SELECT x FROM (VALUES (1), (2)) a (x) FULL JOIN (VALUES (2), (3)) b (x) USING (x)",
+                "VALUES (1, 2, 3)");
```
Comment (haozhun) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: This test is wrong. Should be `VALUES 1, 2, 3` (which is equivalent `VALUES (1), (2), (3)`). Additionally, this test fails before and after fixing this line.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 5]
```diff
@@ -1243,17 +1243,58 @@ public void testJoinUsing()
...
+                + "FROM (\n"
+                + "VALUES (1, 2), (1, 3)) a (x, y)\n"
+                + "JOIN\n"
+                + "(VALUES (1, 2), (1, 3)) b (x, y)\n"
+                + "USING (x, y)\n"
+                + "GROUP BY x\n"
+                + "ORDER BY x",
+                "VALUES (1, 5)");
+
+        MaterializedResult actualLeftJoin = computeActual("SELECT x\n"
```
Comment (haozhun) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: You can do `assertQuery(".....", "VALUES 1, 2"`, which is simpler.

**Comment by haozhun** [comment]
After discussion with the team, we believe we'd like to see this problem fixed in a different way.

In `StatementAnalyzer.visitJoin`, the output `RelationType` should be changed so that only 1 output column is produced for the `using` field.

Given this approach is quite different, I'll close it for now. If you are still interested in fixing this issue, feel free to reopen.

**Comment by yuananf** [comment]
That's ok, maybe I'll look at it when I'm free.

**Comment by haozhun** [comment]
@yuananf Thank you for your understanding

**Pull request closed without merging**